nodeJS/express/controllers.md

```diff
@@ -0,0 +1,466 @@
+
+```javascript
+// user controller file - controllers/userController.js
+
+const asyncHandler = require("express-async-handler");
+
+const getUserById = asyncHandler(async (req, res) => {
+  const userId = req.params.id;
+
+  const user = await someDBQueryToGetUser(userId);
+
+  if (!user) {
+    throw new CustomNotFoundError("User not found");
+  }
+
+  res.send(`User found: ${user.name}`);
+});
+
+const getUsers = asyncHandler(async (req, res) => {
+  // code
+});
+
+const createUser = asyncHandler(async (req, res) => {
+  // code
+});
+
+module.exports = { getUserById, getUsers, createUser };
+```
+
+Route file for the "User"
+
+```javascript
+// route file - routes/userRoutes.js
+
+const express = require('express');
+const userController = require('<path-to-user-controller>');
+
+const router = express.Router();
+
+// You can for example add a top level middleware function that handles say authentication and only let the request come in if they're authenticated
+// This prevents from executing the middleware functions below if the request is not authenticated
+// We will learn more about authentication in later lessons
+// usually calls either next() or next(error)
+
+// router.use(authMiddleware);
+
+// router-level midlewares
+
+// GET request for getting all the users
+router.get('/', userController.getUsers);
+
+// You will likely place your validation/authentication middleware functions here or perhaps in the controller file, e.g.
+// router.post(validationMiddleware, userController.createUser)
+
+// POST request for creating a user
+router.post('/', userController.createUser);
+
+// GET request for getting the user by id
+router.get('/:id', userControoler.getUserById);
+
+module.exports = router;
+```
+
+The main application file, our entry point.
+
+```javascript
+// app file
+
+const express = require('express');
+const morgan = require('morgan');
+const userRouter = require('<path-to-user-routes>');
+
+const app = express();
+
+// application-level middlewares, will always execute on every incoming requests, base mount path is `/`
+
+// a nice to have logger, will automatically show logs for every http request
+app.use(morgan('dev'));
+
+// parses incoming request's body with JSON payloads(e.g. payloads sent through the network with a JSON.stringify'd body in a fetch call as one example)
+// to make it available for use in `req.body`.
+app.use(express.json());
+
+// same as above but parses incoming requests with URL-encoded payloads (e.g name=Damon&role=Professional)
+// more information on extended in the express documentation
+app.use(express.urlencoded({ extended: false }));
+
+app.use((req, res, next) => {
+  // You can of course also create your own for your own use-case!
+  // Just make sure to call `next`
+  next();
+})
+
+// base mount path is `/users` and will always execute on that specific mount path, and yes including `/users/a/b/c`
+app.use('/users', userRouter);
+```
```

Comment:
to "conplete" the code example, can we add an `app.listen` function

Reply:
done